### PR TITLE
add CVE-2022-36804 Detector

### PR DESCRIPTION
**Pull request by SuperX-SIR**
Hey,

this PR for the Vuln Detector Plugin for issue: https://github.com/google/tsunami-security-scanner-plugins/issues/266

prepare the vulnerability enviroment

1.  download image and run the container

use the official docker image

```
docker pull atlassian/bitbucket:8.3.0

docker run  --name="tsunami-bitbucket-8-3" -d -p 37990:7990 -p 37999:7999 atlassian/bitbucket:8.3.0
```

2.  install bitbucket and fill license code

Installing bitbucket requires a license code. After using docker to start the bitbucket environment, you need to access port 37990 and fill in your test license code.Log in to the URL "[my.atlassian.com](http://my.atlassian.com/)" to generate a trial license code.

3.  setup administrator account and create a public repo

- create administrator and login admin:123456
- create a project, name public or whatever
- create a repository or import repositories, just as a example import from git and fill the Clone URL : https://github.com/google/tsunami-security-scanner-plugins
- In Repository settings -> Repository permissions -> public access , enable allow users without Bitbucket account to clone and browse this repository

### CODE REVIEW

**Comment by nttran8** [comment]
Hi @SuperX-SIR! Please run https://github.com/google/google-java-format against your Java files before starting the code review, this would greatly reduce review overhead due to linter errors. Thanks! :)

**Comment by tooryx** [comment]
Hi @SuperX-SIR,

I see that there was no activity on this PR for a while. Are you still willing to contribute?
Otherwise I will close this PR.

_~tooryx_

**Comment by SuperX-SIR** [comment]
Hi，It's stuck in the state：
 Pending — Waiting for internal safe review approval 
Should I do something

**Comment by SuperX-SIR** [comment]
> just wanted to check that you were still willing to contribute.
> I will try to review this PR as soon as possible.

thanks

File: community/detectors/bitbucket_cve_2022_36804/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+[CVE-2022-36804](https://github.com/advisories/GHSA-vcm2-j8f4-m7fj)
+
+## Description of Vulnerability
+
+This detector checks for BitBucket CVE-2022-36804 command injection vulnerability.
+
+A vulnerability in Bitbucket allows a remote, An attacker with access
+to a public Bitbucket repository or with read permissions to a
+private one can execute arbitrary code by sending a malicious
+HTTP request. This All versions released after 6.10.17
```
Comment (tooryx) [inline community/detectors/bitbucket_cve_2022_36804/README.md:14]: Remove "This" at the beginning of the sentence

File: community/detectors/bitbucket_cve_2022_36804/src/test/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804DetectorWithCallbackServerTest.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+                                                                + "HTTP request. This All versions released after 6.10.17 "
+                                                                + "including 7.0.0 and newer are affected, this means that all "
+                                                                + "instances that are running any versions between 7.0.0 and "
+                                                                + "8.3.0 inclusive can be exploited by this vulnerability.")
+                                                .setRecommendation("Upgrade bitbucket to the latest version"))
+                                .build());
+    }
+
+    @Test
+    public void detect_whennoVulnerable_returnsnoVulnerability() throws IOException {
```
Comment (tooryx) [inline community/detectors/bitbucket_cve_2022_36804/src/test/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804DetectorWithCallbackServerTest.java:152]: whennoVulnerable -> whenNoVulnerable

File: community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+                PayloadGeneratorConfig.newBuilder()
+                        .setVulnerabilityType(PayloadGeneratorConfig.VulnerabilityType.BLIND_RCE)
+                        .setInterpretationEnvironment(
+                                PayloadGeneratorConfig.InterpretationEnvironment.LINUX_SHELL)
+                        .setExecutionEnvironment(
+                                PayloadGeneratorConfig.ExecutionEnvironment.EXEC_INTERPRETATION_ENVIRONMENT)
+                        .build();
+
+        if (!payloadGenerator.isCallbackServerEnabled()) {
+            return false;
```
Comment (tooryx) [inline community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java:103]: Please consider adding logging here to inform the user that the plugin requires callback server.

Comment (tooryx) [inline community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java:103]: Please consider adding logging here to inform the user that the plugin requires callback server.

File: community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        name = "Cve202236804VulnDetector",
+        version = "0.1",
+        description =
+                "A vulnerability in Bitbucket allows a remote, An attacker with access "
+                        + "to a public Bitbucket repository or with read permissions to a"
+                        + "private one can execute arbitrary code by sending a malicious "
+                        + "HTTP request. This All versions released after 6.10.17 "
+                        + "including 7.0.0 and newer are affected, this means that all "
+                        + "instances that are running any versions between 7.0.0 and "
+                        + "8.3.0 inclusive can be exploited by this vulnerability.",
```
Comment (tooryx) [inline community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java:51]: Please update the description according to the changes you are doing in the README from my other comments.

File: community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                        Vulnerability.newBuilder()
+                                .setMainId(
+                                        VulnerabilityId.newBuilder()
+                                                .setPublisher("TSUNAMI_COMMUNITY")
+                                                .setValue("CVE_2022_36804"))
+                                .setSeverity(Severity.CRITICAL)
+                                .setTitle(
+                                        "CVE-2022-36804: Bitbucket Command injection vulnerability")
+                                .setDescription(
+                                        "A vulnerability in Bitbucket allows a remote, An attacker with access "
```
Comment (tooryx) [inline community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java:189]: Please change according to the changes you are doing to the README from my previous comments

File: community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                                                + "to a public Bitbucket repository or with read permissions to a"
+                                                + "private one can execute arbitrary code by sending a malicious "
+                                                + "HTTP request. This All versions released after 6.10.17 "
+                                                + "including 7.0.0 and newer are affected, this means that all "
+                                                + "instances that are running any versions between 7.0.0 and "
+                                                + "8.3.0 inclusive can be exploited by this vulnerability.")
+                                .setRecommendation("Upgrade bitbucket to the latest version"))
```
Comment (tooryx) [inline community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java:196]: Please provide the versions that are safe

File: community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+    private DetectionReport buildDetectionReport(
+            TargetInfo targetInfo, NetworkService vulnerableNetworkService) {
+        return DetectionReport.newBuilder()
+                .setTargetInfo(targetInfo)
+                .setNetworkService(vulnerableNetworkService)
+                .setDetectionTimestamp(Timestamps.fromMillis(Instant.now(utcClock).toEpochMilli()))
+                .setDetectionStatus(DetectionStatus.VULNERABILITY_VERIFIED)
+                .setVulnerability(
+                        Vulnerability.newBuilder()
```
Comment (tooryx) [inline community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java:180]: Please also populate the related reference field ([like the xwiki plugin for example](https://github.com/google/tsunami-security-scanner-plugins/blob/3ee364f672d9a567cade29c5399bad493b68d5c9/google/detectors/rce/xwiki/cve202431982/src/main/java/com/google/tsunami/plugins/detectors/rce/cve202431982/Cve202431982Detector.java#L126))

File: community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+                .build();
+    }
+
+    @Override
+    public DetectionReportList detect(TargetInfo targetInfo, ImmutableList<NetworkService> matchedServices) {
+        logger.atInfo().log("Cve202236804VulnDetector starts detecting.");
+        return DetectionReportList.newBuilder()
+                .addAllDetectionReports(
+                        matchedServices.stream()
+                                .filter(Cve202236804VulnDetector::isWebServiceOrUnknownService)
```
Comment (tooryx) [inline community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java:206]: please use `NetworkServiceUtils::isWebService` instead

File: community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+                        .build();
+
+        if (!payloadGenerator.isCallbackServerEnabled()) {
+            return false;
+        }
+        Payload payload = this.payloadGenerator.generate(config);
+
+        String commandToInject = String.format("sh -c \"%s\"", payload.getPayload());
+
+        String PubRepUrl = buildTargetUrl(networkService, GET_ALL_PUB_PATH);
```
Comment (tooryx) [inline community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java:109]: It should be safe to use `NetworkServiceUtils.buildWebApplicationRootUrl(networkService)` directly

File: community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+
+        try {
+            HttpResponse httpResponse =
+                    httpClient.send(get(PubRepUrl).withEmptyHeaders().build(), networkService);
+            if (httpResponse.status().code() == 200
```
Comment (tooryx) [inline community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java:114]: Please prefer an early return approach to make the code more readable. In other words:

if (code != 200) || !contains(STRING_PUB_REP) {
  return false;
}

// continue detection

File: community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+                    && httpResponse.bodyString().get().contains(STRING_PUB_REP)) {
+                String Publink = getArchiveLink(getPubLink(String.valueOf(httpResponse.bodyString())), URLEncoder.encode(commandToInject));
+                if (Publink.length() == 0) {
+                    return false;
+                } else {
```
Comment (tooryx) [inline community/detectors/bitbucket_cve_2022_36804/src/main/java/com/google/tsunami/plugins/detectors/bitbucket/Cve202236804VulnDetector.java:119]: The else is not needed here, given that the previous block will return if it matches

**Comment by tooryx** [comment]
I just tested the plugin that seems to work fine. I would recommend adding a few information to the readme though:

- At least one public repository must exist;
- This repository must already contain some files (must not be uninitialized);
- The repository must have a default branch configured.

_~tooryx_

**Review by tooryx — APPROVED** [review]

**Comment by tooryx** [comment]
Thank you @SuperX-SIR, the PR looks good on my side now (I made some modification internally). It should be merged in a few days.

Could you please also add the instructions to create the test environment in our other repository: https://github.com/google/security-testbeds

Thank you,
_~tooryx_

**Comment by SuperX-SIR** [comment]
> Thank you @SuperX-SIR, the PR looks good on my side now (I made some modification internally). It should be merged in a few days.
> 
> Could you please also add the instructions to create the test environment in our other repository: https://github.com/google/security-testbeds
> 
> Thank you, _~tooryx_

No problem ,I have add a instruction  https://github.com/google/security-testbeds/pull/82

**Pull request merged**